README
NAME
    Class::Dot::Model - Simple way of defining models for DBIx::Class.

VERSION
    This document describes Class::Dot::Model version v%%VERSION%%

SYNOPSIS
         package My::Model;
         use Class::Dot::Model   domain => 'org.mydomain.myapp';
 
         package My::Model::Cat;
         use Class::Dot::Model::Table qw(:has_many);
 
         Table       'cats';
         Columns     qw( id gender dna action colour );
         Primary_Key 'id';
         Has_Many    'memories'
             => 'My::Model::Cat::Memory';
 
         package My::Model::Cat::Memory;
         use Class::Dot::Model::Table qw(:belongs_to);
 
         Table       'memory';
         Columns     qw( id cat content );
         Primary_Key 'id';
         Belongs_To  'cat'
             => 'My::Model::Cat';

    Then you would have to initialize the database configuration for
    org.mydomain.myapp:

         use Config::PlConfig;
         my $DOMAIN   = 'org.mydomain.myapp';
 
         my $dbconfig = {
             driver      => 'mysql',
             database    => 'myappdb',
             hostname    => 'localhost',
             username    => 'me',
             password    => 'secret',
         };
 
         my $plconfig = Config::PlConfig->new({
             domain  => $DOMAIN,
         });
         my $config = $plconfig->load();
         $config->{database} = $dbconfig;
         $plconfig->save();

DESCRIPTION
    A module for making DBIx::Class even simpler.

SUBROUTINES/METHODS
    None of this modules functions should be used directly.

  PRIVATE CLASS METHODS
   "_create_dsn_with_config($config)"
    Create a DBI DSN out of a configuration hash.

   "_install_constructor($call_class, $dsn, $config)"
    Create an install a C<new> function into the callers namespace. This new
    function is responsible for connecting to the database etc.

    The property is a string.

   "_load_dbconfig($config_domain, $config_host)"
    Load the database configuration using a Config::PlConfig host and
    domain.

  PRIVATE SUBROUTINES
   "requires()"
    Used by L<Class::Dot::Model::Preload> to know which modules this module
    requires.

DIAGNOSTICS
    None.

CONFIGURATION AND ENVIRONMENT
    This module uses Config::PlConfig to load configuration if a domain is
    given as import argument.

DEPENDENCIES
    *   DBIx::Class

    *   Class::Dot

    *   Class::Plugin::Util

    *   Params::Util

    *   Config::PlConfig

    *   version

INCOMPATIBILITIES
    None known.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests to
    bug-class-dot-model@rt.cpan.org, or through the web interface at CPAN
    Bug tracker.

SEE ALSO
  DBIx::Class
  Class::Dot::Model::Table
  Class::Dot
AUTHOR
    Ask Solem, ask@0x61736b.net.

LICENSE AND COPYRIGHT
    Copyright (c), 2007 Ask Solem ask@0x61736b.net.

    All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.6 or, at
    your option, any later version of Perl 5 you may have available.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.

    # Local Variables: # mode: cperl # cperl-indent-level: 4 # fill-column:
    78 # End: # vim: expandtab tabstop=4 shiftwidth=4 shiftround



Changes
Revision history for Class-Dot-Model

0.0.1  Sun Sep  9 02:22:58 2007
       Initial release.



t/CatX.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package CatX;

use strict;
use warnings;

use Class::Dot::Model domain => 'org.0x61736b.class.dot.model.test.CatX';

1;

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


t/CatX/Cat.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package CatX::Cat;

use strict;
use warnings;

use Class::Dot::Model::Table qw(:has_many);

Table       'cats';
Columns     qw( id gender dna action colour );
Primary_Key 'id';
Has_Many    'memories'
    => 'CatX::Cat::Memory';

1;

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


t/CatX/Cat/Memory.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package CatX::Cat::Memory;

use strict;
use warnings;

use Class::Dot::Model::Table qw(:belongs_to);

Table           'memory';
Columns         qw(id cat content);
Primary_Key     'id';
Belongs_To      'cat'
    => 'CatX::Cat';

1;

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


Build.PL
# $Id: Build.PL,v 1.5 2007/07/13 00:00:11 ask Exp $
# $Source: /opt/CVS/Getopt-LL/Build.PL,v $
# $Author: ask $
# $HeadURL$
# $Revision: 1.5 $
# $Date: 2007/07/13 00:00:11 $
use strict;
use warnings;

sub BEGIN {
    use File::Spec;
    use FindBin qw($Bin);
    my $inc = File::Spec->catdir($Bin, 'inc');
    unshift @INC, $inc;
}

use Module::Build::Class::Dot::Model;

# ##########################################################



sub YES { 1 };
sub NO  { 0 };

$ENV{DEVEL_COVER_OPTIONS} = '-ignore,blib/Class/InsideOut/Policy/Modwheel.pm';
my $builder = Module::Build::Class::Dot::Model->new(
    module_name          => 'Class::Dot::Model',
    dist_version        => '0.1.2',
    license              => 'perl',
    dist_author          => 'Ask Solem <ASKSH@cpan.org>',
    all_from             => 'lib/Class/Dot/Model.pm',
    dynamic_config       => NO,
    sign                 => NO, # asksh: have to find out why my signature fails.
    create_makefile_pl   => 'traditional',
    create_readme        => YES,
    recursive_test_files => YES,
    requires             => {
        'version'                               => 0,
        'Getopt::LL'                            => 0,
        'Class::Dot'                            => 0,
        'Class::Plugin::Util'                   => 0,
        'Params::Util'                          => 0,
        'Config::PlConfig'                      => 0,
        'DBIx::Class'                           => 0,
    },
    recommends           => {
        'Test::Exception'       => 0.25,
        'IO::Capture::Stderr'   => 0,
    },
    build_requires       => {
     'Test::Simple'            => 0.42,    # 
    },
    add_to_cleanup       => [ qw(
        a.out
        test.pl
        test.c
        test.cpp
        test.m
        *.swp
        .gdb_history
        install.cache
        t/cache
        cache/
    ) ],
    meta_merge => {
        distribution_type   => 'Module',
    },
);

$builder->create_build_script( );

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


lib/Class/Dot/Model/Util.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package Class::Dot::Model::Util;

use strict;
use warnings;
use version; our $VERSION = qv('0.1.2');
use 5.006_001;

use Carp                qw(croak);
use English             qw( -no_match_vars );
use Params::Util        qw(_ARRAY _HASHLIKE);
use Class::Plugin::Util;

my $UPLEVEL  = 2;

my %EXPORT   = (
    FQDN               => \&FQDN,
    install_coderef    => \&install_coderef,
    push_base_class    => \&push_base_class,
    run_as_call_class  => \&run_as_call_class,
);

sub import {
    my $class     = shift;
    my $call_class = caller 0;
    
    while (my ($sub_name, $sub_code_ref) = each %EXPORT) {
        install_coderef(
            $sub_code_ref => $call_class, $sub_name
        );
    }

    return;
}

sub FQDN {
    return join q{::}, @_;
}

sub install_coderef {
    my ($coderef, $package, $name) = @_;

    my $fqdn = join q{::}, $package, $name;

    NOSTRICT: {
        no strict 'refs'; ## no critic;
        *{ FQDN($package, $name) } = $coderef;
    }

    return $fqdn;
}


sub push_base_class {
    my ($base_class, $target_class) = @_;

    Class::Plugin::Util::_require_class($base_class); ## no critic

    if (! $target_class->isa($base_class)) {
        no strict 'refs'; ## no critic
        push @{ FQDN($target_class, 'ISA') }, $base_class;
    }

    run_as_call_class({
        call_class => $target_class,
        class     => $base_class,
        method    => 'import'
    });

    return;
}

sub run_as_call_class {
    my $class;
    my $method;
    my $call_class;

    if (_HASHLIKE($_[0])) {
        my $opts      = shift;
        $class        = $opts->{class};
        $method       = $opts->{method};
        $call_class   = $opts->{call_class};
    }
    else {
        $call_class   = shift;
        $method       = shift;
        $class        = $call_class;
    }

    my $statement = qq{
        package $call_class;
        \$call_class->\$method(\@_);
    };

    eval qq{ $statement }; ## no critic

    if ($EVAL_ERROR) {
        no warnings 'once'; ## no critic
        $Carp::CallLevel = $UPLEVEL; ## no critic
        croak $EVAL_ERROR;
    }

    return;
}

1;

__END__

=begin wikidoc

= NAME

Class::Dot::Model::Util - Private utility functions.

= VERSION

This document describes Class::Dot::Model version v%%VERSION%%

= SYNOPSIS

    # No user servicable parts inside...


= DESCRIPTION

No user serviceable parts inside.

= SUBROUTINES/METHODS

== SUBROUTINES

=== {install_coderef($coderef => $into_class, $method_name)}

Install subroutine into a class with custom name.

Example:

    use Class::Dot::Model qw(install_coderef);
    install_coderef(sub { print "hello world" }, "Hello::World", "hello");

    # prints: "hello world".
    Hello::World->hello();

=== {push_base_class($the_class, $class_target)}

Push a base class into a target class.

Example:

    push_base_class('DBIx::Class',  'MyApp::Model');

Is the same as:

    push @MyApp::Model::ISA, 'DBIx::Class';

    {
        package MyApp::Model;
        DBIx::Class->import();
    }
    

=== {run_as_call_class(($callclass, $method, @arguments) | %options_ref)}

Run a method in a class, but make it look like another class runs it.

Example:

    package Simpsons::Homer;

    sub doh {
        my $caller = caller 0;
        print "$caller: Doh!\n";
    }

    run_as_call_class({
        callclass => 'Hello::World',    - the class to take we're in.
        class     => 'Simpsons::Homer',  - the class that has the method to call.
        method    => 'hello'            - the method to call.
    });

Gives the output:

    Hello::World: Doh!

Because it's basicly the same as writing:

    package Simpsons::Homer;
    
    sub doh {
        my $caller = caller 0;
        print "$caller: Doh!\n";
    }

    package Hello::World;
    Simpsons::Homer->doh();

    package Simpsons::Homer;

    # resume other work.
    
   
There is also a shortcut method of calling this function if you want to run a
method in the callclass itself:

    package Hello::World;
    run_as_callclass('Simpsons::Homer', 'doh');

Gives the output:

    Simpsons::Homer: Doh!


=== {FQDN(@class_components)}

Return a valid Perl class name out of a list of class components.

Example:

    FQDN('Class', 'Plugin', 'Util')

Becomes

    Class::Plugin::Util

= DIAGNOSTICS

None.

= CONFIGURATION AND ENVIRONMENT

This module uses no external configuration or environment variables.

= DEPENDENCIES

* [DBIx::Class]

* [Class::Dot]

* [Class::Plugin::Util]

* [Params::Util]

* [Config::PlConfig]

* [version]

= INCOMPATIBILITIES

None known.

= BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
[bug-class-dot-model@rt.cpan.org|mailto:class-dot-model@rt.cpan.org], or through the web interface at
[CPAN Bug tracker|http://rt.cpan.org].

= SEE ALSO

== [Class::Dot::Model]

== [DBIx::Class]

== [Class::Dot]

== [DBIx::Class::Relationships]

== [DBIx::Class::Manual::Cookbook]

= AUTHOR

Ask Solem, [ask@0x61736b.net].

= LICENSE AND COPYRIGHT

Copyright (c), 2007 Ask Solem [ask@0x61736b.net|mailto:ask@0x61736b.net].

All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

= DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=end wikidoc

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


lib/Class/Dot/Model/Preload.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package Class::Dot::Model::Preload;

use strict;
use warnings;
use version; our $VERSION = qv('0.1.2');
use 5.006_001;

use Class::Dot;
use Params::Util;
use Class::Plugin::Util qw(require_class);

my @PRELOAD_CLASSES = qw(
   Class::Dot::Model 
   Class::Dot::Model::Table 
   Class::Dot::Model::Util
);

sub import {

    for my $class (@PRELOAD_CLASSES) {
        require_class($class);
        if ($class->can('requires')) {
            for my $require_class ($class->requires) {
                require_class($require_class);
            }
        }
    }

    return;
}

1;

__END__

=begin wikidoc

= NAME

Class::Dot::Model::Preload - Preload Class::Dot::Model related modules.

= VERSION

This document describes Class::Dot::Model version v%%VERSION%%

= SYNOPSIS

   # in i.e your startup.pl file:
   #
   use Class::Dot::Model::Preload;

    

= DESCRIPTION

Preload Class::Dot::Model related modules in i.e a mod_perl environment.

= SUBROUTINES/METHODS

This module has no subroutines/methods.


= DIAGNOSTICS

None.

= CONFIGURATION AND ENVIRONMENT

None.

= DEPENDENCIES

* [DBIx::Class]

* [Class::Dot]

* [Class::Plugin::Util]

* [Params::Util]

* [Config::PlConfig]

* [version]

= INCOMPATIBILITIES

None known.

= BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
[bug-class-dot-model@rt.cpan.org|mailto:class-dot-model@rt.cpan.org], or through the web interface at
[CPAN Bug tracker|http://rt.cpan.org].

= SEE ALSO

== [Class::Dot::Model]

= AUTHOR

Ask Solem, [ask@0x61736b.net].

= LICENSE AND COPYRIGHT

Copyright (c), 2007 Ask Solem [ask@0x61736b.net|mailto:ask@0x61736b.net].

All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

= DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=end wikidoc


# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround


lib/Class/Dot/Model.pm
# $Id$
# $Source$
# $Author$
# $HeadURL$
# $Revision$
# $Date$
package Class::Dot::Model;

use strict;
use warnings;
use version; our $VERSION = qv('0.1.2');
use 5.006_001;

use Carp                    qw(croak);
use Params::Util            qw(_ARRAY);
use Config::PlConfig;
use Class::Dot::Model::Util qw(
    push_base_class
    install_coderef run_as_call_class
);

my $BASE_CLASS   = 'DBIx::Class::Schema';

my @INIT_METHODS = qw(
    load_classes
);

my %DSN_PARAMS = (
    hostname    => q{%s},
    database    => q{%s},
    port        => q{%d},
);

my %DSN_DRIVER_REWRITE = (
    mysql => {
        hostname    => 'host',
    },
    Pg    => {
        hostname    => 'hostname',
        database    => 'dbname',
    },
    SQLite => {
        database    => 'dbname',
    },
);

my $dsn_param_rewrite = sub {
    $DSN_DRIVER_REWRITE{$_[0]} ? $DSN_DRIVER_REWRITE{$_[0]} : $_[0];
};

sub requires {
    return $BASE_CLASS;
}

my @MODULES_THAT_IMPORTS_US;

sub import {
    my $class     = shift;
    my $call_class = caller 0;
    my %argv;
   
    if (scalar @_ > 1) {
        %argv = @_;
    }

    push_base_class( $BASE_CLASS => $call_class );

    for my $init_method (@INIT_METHODS) {
        run_as_call_class( $call_class, $init_method );
    }

    return if not defined $argv{domain};
    return if $call_class->can('new');

    my $dbconfig  = _load_dbconfig( $argv{domain}, $argv{host} );
    my $dsnstring = _create_dsn_with_config($dbconfig);

    _install_constructor($call_class, $dsnstring, $dbconfig);

    push @MODULES_THAT_IMPORTS_US, $call_class;

    return;
}

# Install a sighandler that walks through all modules that imports
# us + inherits from DBIx::Class and then disconnects them.
CHECK {
    $SIG{INT} = sub {
        for my $module (@MODULES_THAT_IMPORTS_US) {
            if ($module->isa($BASE_CLASS)) {
                $module->storage->disconnect();
            }
        }
    }
}

sub _load_dbconfig {
    my ($domain, $host) = @_;

    my $plconfig = Config::PlConfig->new({
        host   => $host,
        domain => $domain,
    });
    my $config = $plconfig->load()->{database};

    return $config;
}

sub _install_constructor {
    my ($call_class, $dsn, $config) = @_;

    my $new_coderef = sub {
        my ($class, $options_ref) = @_;
           $options_ref ||= { };

        my $self = $class->connect($dsn, $config->{username}, $config->{password});

        NOSTRICT: {
            no strict 'refs'; ## no critic
            if (my $build_ref = *{ $class . '::BUILD' }{CODE}) { ## no critic
                $build_ref->($self, $options_ref);
            }
        };

        return $self;
    };
        
    return install_coderef($new_coderef => $call_class, 'new');
}

sub _create_dsn_with_config {
    my ($config) = @_;

    # Format the DSN string.
    # %DSN_PARAMS holds the values we support, and values submitted
    # are copied to %dsn_params;
    my %dsn_params;
    my $dsn_format = qq{
        DBI:$config->{driver}:
    };

    DSNPARAM:
    for my $param_name (sort keys %DSN_PARAMS) {
        my $param_type = $DSN_PARAMS{$param_name};
        next DSNPARAM if not defined $config->{$param_name};
        $dsn_params{$param_name} = $config->{$param_name};

        $dsn_format .= join q{=}, (
            $dsn_param_rewrite->($param_name),
            $param_type
        );
        $dsn_format .= q{;};

    };
    chop $dsn_format;
    $dsn_format =~ s/\s*//xmsg;

    no warnings 'uninitialized';  ## no critic;
    my $dsn = sprintf $dsn_format,
        map { $dsn_params{$_} } sort keys %dsn_params;

    return $dsn;
}

1;

__END__

=begin wikidoc

= NAME

Class::Dot::Model - Simple way of defining models for DBIx::Class.

= VERSION

This document describes Class::Dot::Model version v%%VERSION%%

= SYNOPSIS

    package My::Model;
    use Class::Dot::Model   domain => 'org.mydomain.myapp';

    package My::Model::Cat;
    use Class::Dot::Model::Table qw(:has_many);

    Table       'cats';
    Columns     qw( id gender dna action colour );
    Primary_Key 'id';
    Has_Many    'memories'
        => 'My::Model::Cat::Memory';
    
       
    package My::Model::Cat::Memory;
    use Class::Dot::Model::Table qw(:belongs_to);

    Table       'memory';
    Columns     qw( id cat content );
    Primary_Key 'id';
    Belongs_To  'cat'
        => 'My::Model::Cat';

Then you would have to initialize the database configuration for
org.mydomain.myapp:

    use Config::PlConfig;
    my $DOMAIN   = 'org.mydomain.myapp';

    my $dbconfig = {
        driver      => 'mysql',
        database    => 'myappdb',
        hostname    => 'localhost',
        username    => 'me',
        password    => 'secret',
    };

    my $plconfig = Config::PlConfig->new({
        domain  => $DOMAIN,
    });
    my $config = $plconfig->load();
    $config->{database} = $dbconfig;
    $plconfig->save();

= DESCRIPTION

A module for making DBIx::Class even simpler.

= SUBROUTINES/METHODS

None of this modules functions should be used directly.

== PRIVATE CLASS METHODS

=== {_create_dsn_with_config($config)}
=for apidoc string = Class::Dot::Model::_create_dsn_with_config(HASHREF $config)

Create a DBI DSN out of a configuration hash.

=== {_install_constructor($call_class, $dsn, $config)}
=for apidoc CODEREF = Class::Dot::Model::_install_constructor(_CLASS $call_class, string $dsn, HASHREF $config)

Create an install a C<new> function into the callers namespace. This new
function is responsible for connecting to the database etc.

The property is a string.

=== {_load_dbconfig($config_domain, $config_host)}
=for apidoc CODEREF = Class::Dot::Model::_load_dbconfig(string $config_domain, string $config_host)

Load the database configuration using a [Config::PlConfig] host and domain.

== PRIVATE SUBROUTINES

=== {requires()}

Used by L<Class::Dot::Model::Preload> to know which modules this module
requires.

= DIAGNOSTICS

None.

= CONFIGURATION AND ENVIRONMENT

This module uses [Config::PlConfig] to load configuration if a domain is given
as import argument.

= DEPENDENCIES

* [DBIx::Class]

* [Class::Dot]

* [Class::Plugin::Util]

* [Params::Util]

* [Config::PlConfig]

* [version]

= INCOMPATIBILITIES

None known.

= BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
[bug-class-dot-model@rt.cpan.org|mailto:class-dot-model@rt.cpan.org], or through the web interface at
[CPAN Bug tracker|http://rt.cpan.org].

= SEE ALSO

== [DBIx::Class]

== [Class::Dot::Model::Table]

== [Class::Dot]

= AUTHOR

Ask Solem, [ask@0x61736b.net].

= LICENSE AND COPYRIGHT

Copyright (c), 2007 Ask Solem [ask@0x61736b.net|mailto:ask@0x61736b.net].

All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

= DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=end wikidoc


# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround
